src/drivers/intel/cavs/dmic.c

```diff
@@ -1584,6 +1585,102 @@ static int dmic_get_fifo(struct dai *dai, int direction, int stream_id)
 	return dai->plat_data.fifo[SOF_IPC_STREAM_CAPTURE].offset;
 }
 
+/* Functions for HW timestamp */
+
+static inline uint32_t dmic_ts_local_tsctrl_addr(void)
+{
+	return TIMESTAMP_BASE + TS_DMIC_LOCAL_TSCTRL;
+}
+
+static inline uint32_t dmic_ts_local_offs_addr(void)
+{
+	return TIMESTAMP_BASE + TS_DMIC_LOCAL_OFFS;
+}
+
+static inline uint32_t dmic_ts_local_sample_addr(void)
+{
+	return TIMESTAMP_BASE + TS_DMIC_LOCAL_SAMPLE;
+}
+
+static inline uint32_t dmic_ts_local_walclk_addr(void)
+{
+	return TIMESTAMP_BASE + TS_DMIC_LOCAL_WALCLK;
+}
+
+static inline uint32_t dmic_ts_tscc_addr(void)
+{
+	return TIMESTAMP_BASE + TS_DMIC_TSCC;
+}
+
+static int dmic_ts_config(struct dai *dai, struct timestamp_cfg *cfg)
+{
+	if (cfg->type != SOF_DAI_INTEL_DMIC) {
+		trace_dmic_error("dmic_ts_config(): Illegal DAI type");
+		return -EINVAL;
+	}
+
+	cfg->walclk_rate = DMIC_HW_IOCLK;
+
+	return 0;
+}
+
+static int dmic_ts_start(struct dai *dai, struct timestamp_cfg *cfg)
+{
+	uint32_t cdmas;
+	uint32_t addr = dmic_ts_local_tsctrl_addr();
+
+	/* Set DMIC timestamp registers */
+	spin_lock(dai->lock);
+
+	/* First point CDMAS to GPDMA channel that is used by DMIC */
+	cdmas = TS_LOCAL_TSCTRL_CDMAS(cfg->dma_chan_index +
+		cfg->dma_chan_count * cfg->dma_id);
+	io_reg_write(addr, cdmas);
+
+	/* Request on demand timestamp */
+	io_reg_write(addr, TS_LOCAL_TSCTRL_ODTS_BIT | cdmas);
+
+	spin_unlock(dai->lock);
+	return 0;
+}
+
+static int dmic_ts_stop(struct dai *dai, struct timestamp_cfg *cfg)
+{
+	return 0;
```

Comment:
Can we disable the TS, I'm suspicious leaving this ON may cause issues for D3.

Reply:
The ODTS bit is RWS type so it is cleared every time NTK bit is set. But it does not hurt to clear CDMA that is preserved as normal RW type bitfield. CDMAS links TS block to LLP count of the DMA controller/channel so it could cause something.

I'll add to new version an `io_reg_write(dmic_ts_local_tsctrl_addr(), 0);`

Reply:
Done. I added also check into asrc_reset() check of cd->track_drift for executing asrc_dai_stop_timestamp();